extensions/amp-story/1.0/amp-story.js

```diff
@@ -1284,17 +1332,68 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
+    // Enable paywall if required element found in DOM.
+    if (
+      isExperimentOn(this.win, 'amp-story-subscriptions') &&
+      this.element.querySelector('amp-story-subscriptions') != null &&
+      this.storeService_.get(StateProperty.SUBSCRIPTIONS_STATE) ===
+        SubscriptionsState.DISABLED
+    ) {
+      this.storeService_.dispatch(
+        Action.TOGGLE_SUBSCRIPTIONS_STATE,
+        SubscriptionsState.PENDING
+      );
+    }
+
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
+    if (
+      pageIndex >= PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.PENDING
+    ) {
+      return this.blockOnPendingSubscriptionsState_(targetPageId, direction);
+    }
+
+    // Attempt to navigate to the locked pages after the paywall page.
     if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      pageIndex > PAYWALL_PAGE_INDEX &&
+      subscriptionsState === SubscriptionsState.BLOCKED
     ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
+      this.pageAfterGranted_ = targetPageId;
+
+      // Show the paywall immediately if the active page is the paywall page.
```

Comment:
Dont we always show paywall immediately if `pageIndex > PAYWALL_PAGE`. Can we get rid of this check? if we are on paywall page and switchTo paywall page it shouldnt do anything so it should be fine. If we switch to paywall page and UI is already showing, just dont set the 2 sec timer.

Reply:
Nice idea! Indeed by separating out the logic of showing ui and switching to the paywall page we can make everything one-level indentation. Thanks for the advice!